### PR TITLE
multi: support multi-hop payments over private channels

### PR DESCRIPTION
**Pull request by wpaulino**
This PR attempts to implement multi-hop payments over private channels. This is achieved by encoding these channels as hints into an invoice at the time of creation, also known as the `r` field in an invoice in [BOLT-11](https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#tagged-fields). When a node wishes to pay an invoice with this information encoded in it, it will use the hints in order to reach the destination successfully.

### CODE REVIEW

**Comment by Roasbeef** [comment]
> A last_hop flag has been added to addinvoice, which should specify the channel_id of a private channel known to us. This will then look up the edge information from the channel_id and retrieve the information required to be encoded in the invoice.

Haven't yet looked at the code for this PR, but why not just scan the set of channels that the node has inbound, and _automatically_ add routing hints for those _within_ the `AddInvoice` RPC itself?

File: rpcserver.go [hunk 1]
```diff
@@ -2130,13 +2130,41 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 		options = append(options, zpay32.CLTVExpiry(uint64(defaultDelta)))
 	}
 
+	// If some additional routing info is set, we'll make sure to include it
+	// in the invoice as well.
+	if len(invoice.RoutingInfo) > 0 {
```
Comment (Roasbeef) [inline rpcserver.go]: Rather than have the user manually populate the routing info (which seems like quite a burden), we should instead automatically scan our set of private channels, and create routing hints as necessary. This logic would need to traverse the graph one hop from our private channel, to find the _sets_ of channels that the sender can use to successfully route over the last mile of our private channel.

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -1750,6 +1766,93 @@ func addInvoice(ctx *cli.Context) error {
 		return fmt.Errorf("unable to parse receipt: %v", err)
 	}
 
+	switch {
+	case ctx.IsSet("private") && ctx.IsSet("last_hop"):
+		return fmt.Errorf("either `private` or `last_hop` must be " +
+			"set, but not both")
+	case ctx.IsSet("private"):
+		// If the private flag was set, we'll automatically choose a
```
Comment (Roasbeef) [inline cmd/lncli/commands.go]: I expect this logic to actually be within the RPC server. Otherwise, users that are driving `lnd` via the REST or gRPC interface will need to implement this logic manually. In the near future, routing over private channels will be the norm, so I'd prefer to have this logic within `lnd` where we can ensure it's well tested.

Comment (Roasbeef) [reply cmd/lncli/commands.go]: Also we should't select a single channel, and instead should likely max out the routing hint size with all available channels. This is due to the fact that the sender may not actually have paths to the selected channel, and also that the balances of the channels may change even while the invoice is being generated.

File: routing/pathfind.go [hunk 3]
```diff
@@ -617,6 +646,86 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 	return pathEdges, nil
 }
 
+// findPathWithHints is a wrapper around findPath that attempts to find a path
+// to the target with the help of routing hints.
+//
+// NOTE: The hints must be chained together and sorted in forward order in order
+// to reach the target successfully.
+func findPathWithHints(tx *bolt.Tx, graph *channeldb.ChannelGraph,
```
Comment (Roasbeef) [inline routing/pathfind.go]: A set of unit test exercising this new path finding method should be added.

File: routing/pathfind.go [hunk 4]
```diff
@@ -96,6 +101,28 @@ func isSamePath(path1, path2 []*ChannelHop) bool {
 	return true
 }
 
+// Hint is a routing hint that contains the minimum information of a channel
```
Comment (Roasbeef) [inline routing/pathfind.go]: Naming suggestion: Hint -> RoutingHint.

File: routing/pathfind.go [hunk 5]
```diff
@@ -323,7 +350,9 @@ func newRoute(amtToSend lnwire.MilliSatoshi, sourceVertex Vertex,
 		// As a sanity check, we ensure that the selected channel has
 		// enough capacity to forward the required amount which
 		// includes the fee dictated at each hop.
-		if nextHop.AmtToForward.ToSatoshis() > nextHop.Channel.Capacity {
+		if nextHop.Channel.Capacity != unknownCapacity &&
```
Comment (Roasbeef) [inline routing/pathfind.go]: Why's this check necessary? If it's a private channel and we don't yet know the capacity of it?

File: routing/pathfind.go [hunk 6]
```diff
@@ -617,6 +646,86 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
...
+func findPathWithHints(tx *bolt.Tx, graph *channeldb.ChannelGraph,
+	sourceNode *channeldb.LightningNode, target *btcec.PublicKey,
+	hints []Hint, ignoredNodes map[Vertex]struct{},
+	ignoredEdges map[uint64]struct{},
+	amt lnwire.MilliSatoshi) ([]*ChannelHop, error) {
+
+	// If there are no hints available, we can simply find a path to the
+	// destination.
+	if len(hints) == 0 {
+		return findPath(tx, graph, sourceNode, target, ignoredNodes,
```
Comment (Roasbeef) [inline routing/pathfind.go]: Style nit, proper indentation would be: 
```
findPath(
   tx, graph, sourceNode, target, ignoredNodes, ignoredEdges, amt,
)
```

Wrapped to 80 columns as needed.

File: routing/pathfind.go [hunk 7]
```diff
@@ -617,6 +646,86 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
...
+	}
+
+	var finalPath []*ChannelHop
+
+	// First, we'll find a path to the starting node of the first routing
+	// hint. We should be able to do this since the first hint's starting
+	// node should either be advertised to the network, or have a direct
+	// channel with us.
+	publicTarget := hints[0].PubKey
+	path, err := findPath(tx, graph, sourceNode, publicTarget, ignoredNodes,
```
Comment (Roasbeef) [inline routing/pathfind.go]: Similar style nit here.

File: routing/pathfind.go [hunk 8]
```diff
@@ -617,6 +646,86 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the hops from the path found above to the final path.
+	finalPath = append(finalPath, path...)
+
+	// Then, we'll go over our list of hints that should reach the
+	// target destination when chained together.
+	for i, hint := range hints {
```
Comment (Roasbeef) [inline routing/pathfind.go]: This logic seems to assume a single routing hint per invoice, and also that the routing hint spans multiple hops. Instead, due to current protocol limitations, it perhaps should instead treat each routing hint in a discrete manner. As it's possible for the receiver to have many private outgoing channels to distinct nodes. This logic would the be modified to use the first non-blacklisted-by-mission-control routing hint first, then as routes possibly fail, the rest of the routing hints would be filtered out.

File: zpay32/invoice.go [hunk 9]
```diff
@@ -149,24 +149,7 @@ type Invoice struct {
...
-	FeeBaseMsat uint32
-
-	// FeeProportionalMillionths is the proportional fee in millionths of a
-	// satoshi required for routing along this channel.
-	FeeProportionalMillionths uint32
-
-	// CltvExpDelta is this channel's cltv expiry delta.
-	CltvExpDelta uint16
-}
+type ExtraRoutingInfo = routing.Hint
```
Comment (Roasbeef) [inline zpay32/invoice.go]: 👍

File: lnd_test.go [hunk 10]
```diff
@@ -3032,6 +3088,199 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
 	}
 }
 
+// testMultiHopOverPrivateChannels tests that private channels can be used as
+// intermediate hops in a route for payments.
+func testMultiHopOverPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
+	// We'll test that multi-hop payments over private channels work as
+	// intended. To do so, we'll create the following topology:
+	//         public         private          private
+	// Alice <--100k--> Bob <--100k--> Carol <--100k--> Dave
```
Comment (Roasbeef) [inline lnd_test.go]: Related to my earlier comments in the routing package, I'm not sure this is actually yet achievable within the constraints of the protocol. How would Dave obtain information concerning the Bob <-> Carol private channel? Instead, I'd expect the topology to be something like: 
```
//      private                public                  private
// Alice <--100k--> Bob <--100k--> Carol <--100k--> Dave
```

Comment (halseth) [reply lnd_test.go]: The spec is a bit vague, but from what I read you can provide multiple ordered channels (a route), not just single last hops. I always assumed this was to allow cases like for instance Dave being a good friend of Bob, so Bob has shared his channels with him (or maybe Dave can purchase knowledge about private channels from a provider?).

File: rpcserver.go (hunk 1 above)
Comment (wilmerpaulino) [reply rpcserver.go]: Fixed.

File: cmd/lncli/commands.go (hunk 2 above)
Comment (wilmerpaulino) [reply cmd/lncli/commands.go]: Fixed.

File: routing/pathfind.go (hunk 3 above)
Comment (wilmerpaulino) [reply routing/pathfind.go]: Fixed.

File: routing/pathfind.go (hunk 5 above)
Comment (wilmerpaulino) [reply routing/pathfind.go]: This is necessary because the routing hint doesn't include the capacity of the channel, so it defaults to `0` when creating the `channeldb.ChannelEdgePolicy`.

File: routing/pathfind.go (hunk 7 above)
Comment (wilmerpaulino) [reply routing/pathfind.go]: Fixed.

File: routing/pathfind.go (hunk 8 above)
Comment (wilmerpaulino) [reply routing/pathfind.go]: Fixed.

File: lnd_test.go (hunk 10 above)
Comment (wilmerpaulino) [reply lnd_test.go]: Modified the test case to follow the model above.

Something like
```
//         public        private          private
// Alice <--100k--> Bob <--100k--> Carol <--100k--> Dave
```
is still possible with the current implementation as long as Dave knows about the channel between Bob and Carol. I can add another test case for that if needed.

**Comment by wilmerpaulino** [comment]
Pushed out a new version that treats the routing hints as an extension of the channel graph by modifying our path finding algorithm to also use a set of additional edges. I decided to implement it this way to avoid mutating the current channel graph and/or making a copy. Feedback welcomed for other/better approaches.

@Roasbeef PTAL.

File: routing/pathfind.go [hunk 11]
```diff
@@ -596,6 +630,49 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 		if err != nil {
 			return nil, err
 		}
+
+		// Then, we'll examine all the additional edges from the node
+		// we're currently visiting.
+		for _, edge := range additionalEdges[bestNode.PubKeyBytes] {
+			v := Vertex(edge.Node.PubKeyBytes)
```
Comment (halseth) [inline routing/pathfind.go]: Since you are doing the same here and in the loop above, this can be extracted into a helper closure that can be called from both loops:

```go
for _, edge := range additionalEdges[bestNode.PubKeyBytes] {
    capacity := unknownCapacity
    processEdge(edge, capacity)
```

File: routing/pathfind.go (hunk 5 above)
Comment (halseth) [reply routing/pathfind.go]: Since we probably can assume that the hints given to us have enough capacity to carry the payment, can't we just set `capacity = paymentAmt` (+ enough to account for fees) for the routing hints?

File: routing/pathfind.go (hunk 11 above)
Comment (wpaulino) [reply routing/pathfind.go]: Fixed.

File: lnd_test.go [hunk 12]
```diff
@@ -2879,43 +2879,21 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
...
-
-	// One block is enough to make the channel ready for use, since the
-	// nodes have defaultNumConfs=1 set.
-	block := mineBlocks(t, net, 1)[0]
-	chanPointPrivate, err := net.WaitForChannelOpen(ctxb, chanOpenUpdate)
-	if err != nil {
-		t.Fatalf("error while waiting for channel open: %v", err)
-	}
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	chanPointPrivate := openChannelAndAssert(ctxt, t, net, carol, net.Alice,
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lnd_test.go]: Fixed.

File: rpcserver.go [hunk 13]
```diff
@@ -2163,13 +2164,71 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+	// hints for them.
+	if invoice.Private {
+		openChannels, err := r.server.chanDB.FetchAllChannels()
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch all channels")
+		}
+
+		graph := r.server.chanDB.ChannelGraph()
+
+		// TODO: restrict number of route hints?
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply rpcserver.go]: Fixed.

File: rpcserver.go [hunk 14]
```diff
@@ -3077,6 +3102,30 @@ func (r *rpcServer) DecodePayReq(ctx context.Context,
 	// explicitly.
 	expiry := int64(payReq.Expiry().Seconds())
 
+	// Convert between the `lnrpc` and `routing` types.
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply rpcserver.go]: Fixed.

File: zpay32/invoice.go [hunk 15]
```diff
@@ -976,25 +972,29 @@ func writeTaggedFields(bufferBase32 *bytes.Buffer, invoice *Invoice) error {
...
-			binary.BigEndian.PutUint32(base256[45:49], r.FeeProportionalMillionths)
-			binary.BigEndian.PutUint16(base256[49:51], r.CltvExpDelta)
-			routingDataBase256 = append(routingDataBase256, base256...)
+	for _, routeHint := range invoice.RouteHints {
+		// Each hop hint is encoded using 51 bytes, so we'll make to
+		// sure to allocate enough space for the whole route hint.
+		routeHintBase256 := make([]byte, 0, hopHintLen*len(routeHint))
+
+		for _, hopHint := range routeHint {
+			base256 := make([]byte, hopHintLen)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply zpay32/invoice.go]: Fixed.

File: zpay32/invoice.go [hunk 16]
```diff
@@ -849,35 +842,38 @@ func parseFallbackAddr(data []byte, net *chaincfg.Params) (btcutil.Address, erro
...
+	if len(base256Data)%hopHintLen != 0 {
+		return nil, fmt.Errorf("expected length multiple of %d bytes, "+
+			"got %d", hopHintLen, len(base256Data))
 	}
 
-	var routingInfo []ExtraRoutingInfo
-	info := ExtraRoutingInfo{}
+	var routeHint []routing.HopHint
+
+	hopHint := routing.HopHint{}
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply zpay32/invoice.go]: Fixed.

File: routing/pathfind_test.go [hunk 17]
```diff
@@ -489,6 +489,70 @@ func TestBasicGraphPathFinding(t *testing.T) {
...
+		ChannelID:                 1337,
+		FeeBaseMSat:               1,
+		FeeProportionalMillionths: 1000,
+		TimeLockDelta:             9,
+	}
+
+	additionalEdges := make(map[Vertex][]*channeldb.ChannelEdgePolicy)
+	songokuVertex := NewVertex(aliases["songoku"])
+	songokuAdditionalEdges := []*channeldb.ChannelEdgePolicy{songokuToDoge}
+	additionalEdges[songokuVertex] = songokuAdditionalEdges
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply routing/pathfind_test.go]: Fixed.

File: routing/pathfind.go [hunk 18]
```diff
@@ -489,6 +496,33 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
...
+	}
+
+	// We can't always assume that the end destination is publicly
+	// advertised to the network and included in the graph.ForEachNode call
+	// above, so we'll manually include the target node if it's not already
+	// there.
+	targetVertex := NewVertex(target)
+	if _, exists := distance[targetVertex]; !exists {
+		targetNode := &channeldb.LightningNode{}
+		targetNode.AddPubKey(target)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply routing/pathfind.go]: Fixed.

File: routing/pathfind.go [hunk 19]
```diff
@@ -28,6 +28,28 @@ const (
 	infinity = math.MaxInt64
 )
 
+// HopHint is a routing hint that contains the minimum information of a channel
+// required for an intermediate hop in a route to forward the payment to the
+// next. This should be ideally used for private channels, since they are not
+// publicly advertised to the network for routing.
+type HopHint struct {
+	// PublicKey is the public key of the node at the start of the channel.
+	PublicKey *btcec.PublicKey
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply routing/pathfind.go]: Fixed.

File: zpay32/invoice.go [hunk 20]
```diff
@@ -524,10 +513,12 @@ func validateInvoice(invoice *Invoice) error {
 		return fmt.Errorf("neither description nor description hash set")
 	}
 
-	// Can have at most 20 extra hops for routing.
-	if len(invoice.RoutingInfo) > 20 {
-		return fmt.Errorf("too many extra hops: %d",
-			len(invoice.RoutingInfo))
+	// Each route hint can have at most 20 hops.
+	for _, routeHint := range invoice.RouteHints {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply zpay32/invoice.go]: 20 route hints should definitely be plenty when creating the invoice, but do we really want to mark an invoice as invalid just because it has more than that?

File: routing/pathfind.go (hunk 5 above)
Comment (wpaulino) [reply routing/pathfind.go]: Are there any drawbacks to the check included above? It was also useful in [`cc64a54`](https://github.com/lightningnetwork/lnd/pull/846/commits/cc64a54346161890d5c0bb000b89a38ea4251c05#diff-ef7018e489d1a522e140e134eda04fe9R565).

File: zpay32/invoice.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (halseth) [reply zpay32/invoice.go]: We should at least restrict the number when creating invoices. I also think rejecting invoices with clearly excessive number of route hints (maybe this limit should be a bit higher than 20?) is reasonable.

File: routing/pathfind.go [hunk 21]
```diff
@@ -514,15 +514,78 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
...
+	// satisfy our specific requirements.
+	processEdge := func(edge *channeldb.ChannelEdgePolicy,
+		capacity btcutil.Amount, pivot Vertex) {
+
+		v := Vertex(edge.Node.PubKeyBytes)
+
+		// If the edge is currently disabled, then we'll stop here, as
+		// we shouldn't attempt to route through it.
+		edgeFlags := lnwire.ChanUpdateFlag(edge.Flags)
+		if edgeFlags&lnwire.ChanUpdateDisabled == lnwire.ChanUpdateDisabled {
```
Comment (halseth) [inline routing/pathfind.go]: nit: just check `!= 0`

File: rpcserver.go [hunk 22]
```diff
@@ -2181,19 +2181,52 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 
 		graph := r.server.chanDB.ChannelGraph()
 
-		// TODO: restrict number of route hints?
 		var routeHints [][]routing.HopHint
 		for _, channel := range openChannels {
+			// We'll restrict the number of individual route hints
+			// to 20 to avoid creating overly large invoices.
+			if len(routeHints) > 20 {
```
Comment (halseth) [inline rpcserver.go]: 👍

File: rpcserver.go [hunk 23]
```diff
@@ -2181,19 +2181,52 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
 			// we'll skip public ones.
 			isPublic := channel.ChannelFlags&lnwire.FFAnnounceChannel != 0
 			if isPublic {
 				continue
 			}
 
-			chanID := channel.ShortChanID.ToUint64()
+			// Make sure the channel has enough capacity for our
+			// amount.
+			if amt > channel.Capacity {
```
Comment (halseth) [inline rpcserver.go]: As the capacity might not be available for incoming payments, we should instead get the balance available to the remote peer. This doesn't have to be exact, but will at least avoid adding channels having no balance for incoming payments to the routing hints.

File: routing/pathfind.go (hunk 5 above)
Comment (halseth) [reply routing/pathfind.go]: Drawback as I see it is that another constant is introduced even not strictly necessary. If we always assumed the private channel had enough capacity, we wouldn't need the check in https://github.com/lightningnetwork/lnd/pull/846/commits/cc64a54346161890d5c0bb000b89a38ea4251c05#diff-ef7018e489d1a522e140e134eda04fe9R565.

**Comment by Roasbeef** [comment]
Can you rebase this to master? Thanks!

File: routing/pathfind.go (hunk 5 above)
Comment (wpaulino) [reply routing/pathfind.go]: Fixed.

File: zpay32/invoice.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (wpaulino) [reply zpay32/invoice.go]: Fixed.

File: routing/pathfind.go (hunk 21 above)
Comment (wpaulino) [reply routing/pathfind.go]: Fixed.

File: rpcserver.go (hunk 23 above)
Comment (wpaulino) [reply rpcserver.go]: Good point! Fixed.

**Comment by wpaulino** [comment]
Rebased.

**Comment by Roasbeef** [comment]
Travis build looks to be consistently failing: 
```
    --- FAIL: TestLightningNetworkDaemon/multi-hop_payments_over_private_channels (11.18s)
    	lnd_test.go:69: Failed: (multi-hop payments over private channels): exited with error: 
    		*errors.errorString unable to send payments from alice to dave: received payment error: unable to find a path to destination
    		/home/travis/gopath/src/github.com/lightningnetwork/lnd/lnd_test.go:3206 (0xb77587)
```

File: lnrpc/rpc.proto [hunk 24]
```diff
@@ -1442,6 +1442,9 @@ message Invoice {
     invoice's destination.
     */
     repeated RouteHint route_hints = 14 [json_name = "route_hints"];
+
+    /// Whether this invoice should include routing hints for private channels.
+    bool private = 15 [json_name = "private"];
```
Comment (Roasbeef) [inline lnrpc/rpc.proto]: Perhaps we should just make this the default on the `lncli` level? Nice that users are able to toggle it in any case.

File: rpcserver.go [hunk 25]
```diff
@@ -2193,13 +2194,104 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			}
+
+			// Since we're only interested in our private channels,
+			// we'll skip public ones.
+			isPublic := channel.ChannelFlags&lnwire.FFAnnounceChannel != 0
+			if isPublic {
+				continue
+			}
+
+			// Make sure the counterparty has enough balance for our
```
Comment (Roasbeef) [inline rpcserver.go]: counterparty -> channel

Comment (Roasbeef) [reply rpcserver.go]: This is a nice touch, we could even _always_ apply something like this in order to reduce payment latency due to payment failures. Rather than them trying the other 100 channels I have open, they can know to instead only direct a payment to a  _particular_ channel.

File: rpcserver.go [hunk 26]
```diff
@@ -2193,13 +2194,104 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			var peerOnline, linkActive bool
+			_, err = r.server.FindPeer(channel.IdentityPub)
+			if err == nil {
+				peerOnline = true
+			}
+
+			channelID := lnwire.NewChanIDFromOutPoint(
+				&channel.FundingOutpoint,
+			)
+			link, err := r.server.htlcSwitch.GetLink(channelID)
```
Comment (Roasbeef) [inline rpcserver.go]: It's the case that if a channel is in the switch, then the peer is already online. As a result, we can remove the first check in favor of only this one.

File: rpcserver.go [hunk 27]
```diff
@@ -2193,13 +2194,104 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			if err != nil {
+				continue
+			}
+
+			// Now, we'll need to determine which is the correct
+			// policy for HTLCs being sent from the remote node.
+			var remotePolicy *channeldb.ChannelEdgePolicy
+
+			remotePub := channel.IdentityPub.SerializeCompressed()
+			if bytes.Equal(remotePub, p1.Node.PubKeyBytes[:]) {
```
Comment (Roasbeef) [inline rpcserver.go]: This should be `bytes.Compare`, and actually make a decision based on the _ordering_ (not equality) of the two pubkeys.

File: routing/router.go [hunk 28]
```diff
@@ -1482,6 +1482,16 @@ type LightningPayment struct {
...
 	PayAttemptTimeout time.Duration
 
+	// RouteHints represents the different routing hints that can be used to
+	// assist a payment in reaching its destination successfully. These
+	// hints will act as intermediate hops along the route.
+	//
+	// NOTE: This is optional unless required by the payment. When providing
+	// multiple routes, ensure the hop hints within each route are chained
+	// together and sorted in forward order in order to reach the
+	// destination successfully.
```
[the remark this answers is not in the document]
Comment (halseth) [reply routing/router.go]: I think this is because each hint only specifies the _first_ node for the channel, so you must assume the next hint contains the _second_ node.

File: cmd/lncli/commands.go [hunk 29]
```diff
@@ -1709,6 +1709,12 @@ var addInvoiceCommand = cli.Command{
 				"specified an expiry of 3600 seconds (1 hour) " +
 				"is implied.",
 		},
+		cli.BoolTFlag{
```
Comment (halseth) [inline cmd/lncli/commands.go]: TIL about `BoolTFlag`! 😀

File: rpcserver.go [hunk 30]
```diff
@@ -2193,13 +2194,104 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			isActive := peerOnline && linkActive
+			if !isActive {
+				continue
+			}
+
+			// Fetch the policies for each end of the channel.
+			chanID := channel.ShortChanID.ToUint64()
+			_, p1, p2, err := graph.FetchChannelEdgesByID(chanID)
+			if err != nil {
+				continue
```
[the remark this answers is not in the document]
Comment (halseth) [reply rpcserver.go]: No, only for errors useful for debugging failures are enough :)

EDIT: seems like the fixup was added to another commit

File: rpcserver.go [hunk 31]
```diff
@@ -2244,15 +2248,17 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
-			if !isActive {
+			if !linkActive {
 				continue
 			}
 
 			// Fetch the policies for each end of the channel.
 			chanID := channel.ShortChanID.ToUint64()
 			_, p1, p2, err := graph.FetchChannelEdgesByID(chanID)
 			if err != nil {
+				rpcsLog.Debugf("Unable to fetch the routing "+
```
Comment (halseth) [inline rpcserver.go]: this should maybe be a fixup to a different commit? 😛

File: lnrpc/rpc.proto [hunk 32]
```diff
@@ -1353,6 +1353,34 @@ message ClosedChannelUpdate {
     ChannelPoint chan_point = 4;
 }
 
+message HopHint {
+    /// The public key of the node at the start of the channel.
+    string pub_key = 1 [json_name = "pub_key"];
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lnrpc/rpc.proto]: Fixed.

File: lnrpc/rpc.proto [hunk 33]
```diff
@@ -1497,6 +1531,7 @@ message PayReq {
     string description_hash = 7 [json_name = "description_hash"];
     string fallback_addr = 8 [json_name = "fallback_addr"];
     int64 cltv_expiry = 9 [json_name = "cltv_expiry"];
+    repeated RouteHint route_hints = 14 [json_name = "route_hints"];
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply lnrpc/rpc.proto]: Whoops, copy-pasta error. Fixed.

File: rpcserver.go (hunk 31 above)
Comment (wpaulino) [reply rpcserver.go]: Whoops, fixed.

File: lnd_test.go [hunk 34]
```diff
@@ -3131,6 +3131,172 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
...
+
+	resp, err := dave.AddInvoice(ctxb, invoice)
+	if err != nil {
+		t.Fatalf("unable to add invoice for dave: %v", err)
+	}
+
+	// Let Alice pay the invoice.
+	payReqs := []string{resp.PaymentRequest}
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	err = completePaymentRequests(ctxt, net.Alice, payReqs, true)
```
Comment (Roasbeef) [inline lnd_test.go]: There seems to be a race in this test on travis. To resolve it, we should add `WaitForNetWorkChanOpen` calls for all relevant nodes to ensure they have the proper topology before attempting any payments.

File: lnd_test.go [hunk 35]
```diff
@@ -3945,7 +3964,7 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
 	// maxFundingAmount (2^24) satoshis value.
 	ctxt, _ := context.WithTimeout(ctxb, timeout)
 	chanPoint := openChannelAndAssert(
-		ctxt, t, net, dave, carol, chanAmt, pushAmt,
+		ctxt, t, net, net.Alice, carol, chanAmt, pushAmt, false,
```
Comment (halseth) [inline lnd_test.go]: `dave` has been changed to `net.Alice` here. That's probably why Travis is failing.

File: lnd_test.go (hunk 34 above)
Comment (wpaulino) [reply lnd_test.go]: Ah, wasn't sure what was causing that. Thanks for the tip!

File: lnd_test.go (hunk 35 above)
Comment (wpaulino) [reply lnd_test.go]: Nice catch! Not sure how this happened 🤔

**Comment by wpaulino** [comment]
Travis is green now 😎

File: rpcserver.go [hunk 36]
```diff
@@ -2202,13 +2203,105 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+
+			if !linkActive {
+				continue
+			}
+
+			// Fetch the policies for each end of the channel.
+			chanID := channel.ShortChanID.ToUint64()
+			_, p1, p2, err := graph.FetchChannelEdgesByID(chanID)
+			if err != nil {
+				rpcsLog.Debugf("Unable to fetch the routing "+
```
Comment (halseth) [inline rpcserver.go]: nit: should be `Errorf`

File: rpcserver.go [hunk 37]
```diff
@@ -2202,13 +2203,105 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			// channel for our amount. We do this in order to reduce
+			// payment errors when attempting to use this channel
+			// as a hint.
+			if amtMSat >= channel.LocalCommitment.RemoteBalance {
+				continue
+			}
+
+			// Make sure the channel is active. To do so, we'll need
+			// to satisfy the following:
+			//	1. The counterparty is online.
```
Comment (halseth) [inline rpcserver.go]: We only do one check now?

File: rpcserver.go [hunk 38]
```diff
@@ -2202,13 +2203,105 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			if isPublic {
+				continue
+			}
+
+			// Make sure the counterparty has enough balance in the
+			// channel for our amount. We do this in order to reduce
+			// payment errors when attempting to use this channel
+			// as a hint.
+			if amtMSat >= channel.LocalCommitment.RemoteBalance {
+				continue
```
Comment (halseth) [inline rpcserver.go]: Actually, I would add a debug log here as well. This will only happen when people are making invoices, so shouldn't be spamming the logs.

File: rpcserver.go [hunk 39]
```diff
@@ -2202,13 +2209,105 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+					"have enough remote balance to be "+
+					"used as a routing hint for invoice",
+					chanPoint)
+				continue
+			}
+
+			// Make sure the channel is active.
+			var linkActive bool
+			link, err := r.server.htlcSwitch.GetLink(chanPoint)
+			if err == nil {
```
Comment (halseth) [inline rpcserver.go]: Hm, I had a comment about this, not sure why it disappeared ;p

Suggestion:
```go
if err != nil {
    log.Errorf("Unable to get link %v, skipping: %v", chanID, err)
   continue
}
if !link.EligibleToForward() {
    log.Debugf("link %v not eligible, skippping", chanID)
    continue
}
```

Comment (wpaulino) [reply rpcserver.go]: Fixed.

File: cmd/lncli/commands.go (hunk 29 above)
Comment (Roasbeef) [reply cmd/lncli/commands.go]: I don't think we can quite flip this switch yet. We'll need to wait until we know _all_ the other implementations support height hints, and also our implementation. As a result, this should be changed to a regular `Bool` flag, with the default being `false`.

File: rpcserver.go [hunk 40]
```diff
@@ -2202,13 +2209,107 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
...
+			if amtMSat >= channel.LocalCommitment.RemoteBalance {
+				rpcsLog.Debugf("Skipping channel %v due to "+
+					"not having enough remote balance",
+					chanPoint)
+				continue
+			}
+
+			// Make sure the channel is active.
+			link, err := r.server.htlcSwitch.GetLink(chanPoint)
+			if err == nil {
```
Comment (Roasbeef) [inline rpcserver.go]: From my reading this should be `err != nil`, and then a continue. Otherwise, we' can panic below if we create an invoice when all channels are inactive.

File: cmd/lncli/commands.go (hunk 29 above)
Comment (Roasbeef) [reply cmd/lncli/commands.go]: bool af
